packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts

```diff
@@ -13,26 +15,116 @@ function asyncEval(url) {
  * Adds React Native specific tweaks to bootstrap logic.
  */
 export default class ReactNativeEnvPlugin {
+  constructor(private bundle: boolean) {}
+
   apply(compiler: webpack.Compiler) {
-    compiler.hooks.thisCompilation.tap('ReactNativeEnvPlugin', compilation => {
-      (compilation.mainTemplate as any).hooks.bootstrap.tap(
-        'ReactNativeEnvPlugin',
-        (source: string) => {
-          return `${asyncEval}\n${source}`;
+    if (this.bundle) {
+      // When creating static bundle (non-RAM), async chunks will be concatenated into main bundle.
+      // This will allow easy switching between RAM bundle and non-RAM static bundle.
+      compiler.hooks.emit.tap('ReactNativeEnvPlugin', compilation => {
+        // Skip if there is no async chunks.
+        if (compilation.chunks.length === 1) {
+          return;
         }
-      );
+
+        const {
+          mainMap,
+          mainSource: mainSourceFilename,
+          asyncChunks,
+        } = this.getFilenamesFromChunks(compilation.chunks);
+
+        const sourceMappingRegex = /\/\/# sourceMappingURL=(.+)/;
+        const mainSource = compilation.assets[mainSourceFilename].source();
+        const sourceMappingMatch = new RegExp(sourceMappingRegex, 'g').exec(
+          mainSource
+        );
+
+        // Concatenate all chunks and it's source maps. Chunks source needs to have source mapping URL
+        // removed, since it will be added at the end of the whole bundle.
+        const concat = new Concat(true, mainSourceFilename, '\n');
+        concat.add(
+          mainSourceFilename,
+          mainSource.replace(new RegExp(sourceMappingRegex, 'g'), ''),
+          compilation.assets[mainMap].source()
+        );
+        asyncChunks.forEach(chunk => {
+          concat.add(
+            chunk.source,
+            (compilation.assets[chunk.source].source() as string).replace(
+              new RegExp(sourceMappingRegex, 'g'),
+              ''
+            ),
+            compilation.assets[chunk.map].source()
+          );
+        });
+
+        // Add single source mapping url pointing to file with concatenated source maps.
+        if (sourceMappingMatch) {
+          concat.add(null, sourceMappingMatch[0]);
+        }
+
+        // Remove non-main assets
+        Object.keys(compilation.assets)
+          .filter(item => ![mainSourceFilename, mainMap].includes(item))
+          .forEach(item => {
```

Comment:
Not sure how big this array may get, but if it's 1000+ items it's probably worth to have only one forEach and move filtering inside

Reply:
Each `import()` creates new asset + main bundle and main bundle maps. Not sure if someone will even use so many `import()` to go beyond few hundreds.